crates/futures/src/task/multithread.rs

```diff
@@ -0,0 +1,201 @@
+use std::future::Future;
+use std::mem::ManuallyDrop;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::sync::Arc;
+use std::sync::atomic::AtomicI32;
+use std::sync::atomic::Ordering::SeqCst;
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use crate::queue::Queue;
+
+
+#[wasm_bindgen]
+extern "C" {
+    type Atomics;
+
+    #[wasm_bindgen(static_method_of = Atomics, js_name = waitAsync)]
+    fn wait_async(buf: &JsValue, index: i32, value: i32) -> js_sys::Promise;
+
+    #[wasm_bindgen(static_method_of = Atomics, js_name = waitAsync, getter)]
+    fn get_wait_async() -> JsValue;
+}
+
+fn wait_async(ptr: &AtomicI32, current_value: i32) -> js_sys::Promise {
+    // If `Atomics.waitAsync` isn't defined (as it isn't defined anywhere today)
+    // then we use our fallback, otherwise we use the native function.
+    if Atomics::get_wait_async().is_undefined() {
+        crate::task::wait_async_polyfill::wait_async(ptr, current_value)
+
+    } else {
+        let mem = wasm_bindgen::memory().unchecked_into::<js_sys::WebAssembly::Memory>();
+        Atomics::wait_async(&mem.buffer(), ptr as *const AtomicI32 as i32 / 4, current_value)
+    }
+}
+
+fn notify(atomic: &AtomicI32, wakeup: u32) {
+    unsafe {
+        core::arch::wasm32::atomic_notify(
+            // TODO verify the safety of this
```

Comment:
For this the safety here looks fine by me, so this should be safe to remove the comment

Reply:
Since I'm not well versed in unsafe voodoo: why is it safe to cast a `&` to a `*mut`? Doesn't that violate Rust's memory aliasing?

Reply:
`&AtomicI32` has interior mutability through its usage of `UnsafeCell` and while it'd be best to have an inherent method to go from `&AtomicI32` to `*mut i32` this is basically the same. Note that `&` doesn't mean "readonly" it just means "shared"

Reply:
Yeah, but I mean, can't Rust make assumptions that the `*mut` doesn't alias (since it's `*mut`), and then make optimizations based on that, but then that will break since it's actually aliased?

Reply:
I won't pretend to really understand all the various aspects of the unsafe guidelines, but this pattern is basically required to work in one way or another.